internal/command/cliconfig/provider_installation.go

```diff
@@ -253,7 +254,8 @@ func decodeProviderInstallationFromConfig(hclFile *hclast.File) ([]*ProviderInst
 						))
 						continue
 					}
-					dirPath := filepath.Clean(rawPath)
+					interpolatedPath := os.ExpandEnv(rawPath)
```

Comment:
Suppose some `${MYVAR}` happens to get unset. `os.ExpandEnv` will then provide an empty string as the value, which then gets removed with `filepath.Clean()`. So if the raw path is `foo/${MYVAR}/path/to/overrides`, we end up reading `foo/path/to/overrides` which inevitably results in an error downstream once `terraform init` validates the provider paths. Or in a more nefarious scenario, it loads some malicious binary. 

One QoL improvement is to use `os.Expand()` instead so we can properly render a warning diagnostic when this occurs. 
```go
interpolatedPath := os.Expand(rawPath, func(placeholder string) string {
  if v, ok := os.LookupEnv(placeholder); ok {
    return v
  } else {
    diags = diags.Append(tfdiags.Sourceless(
      tfdiags.Warning,
      "Interpolated environment variable not set",
      fmt.Sprintf("The environment variable %s is not set or empty and can result in undesired behavior", placeholder),
    ))
    return ""
  }
})
```
If we feel strongly we should error instead of providing a warning at this point, I'm all for it.

Reply:
Thanks for the recommendation @sebasslash, this makes sense!

I ended up issuing an error instead of warning:

```hcl
provider_installation {                      
  dev_overrides {                            
    "craigsloggett/github" = "${NOT_SET}/bin"
  }                                          
                                             
  direct {}                                  
}                                            
```

```shell
~/.cache/go/bin $ ./terraform --version
There are some problems with the CLI configuration:
╷
│ Error: Interpolated environment variable not set
│
│ The environment variable NOT_SET is not set or empty and can result in
│ undesired behavior
╵

As a result of the above problems, Terraform may not behave as intended.


Terraform v1.9.0-dev
on darwin_arm64
```

Reply:
After doing some testing, I've found that multiple uses of the same environment variable in the string, if unset, would result in multiple identical error/warning diagnostics.

For example, this configuration:

```hcl
provider_installation {                      
  dev_overrides {                            
    "craigsloggett/github" = "${NOT_SET}/${NOT_SET}/bin"
  }                                          
                                             
  direct {}                                  
}       
```

Would result in this output:

```shell
terraform --version
There are some problems with the CLI configuration:
╷
│ Error: Interpolated environment variable not set
│
│ The environment variable NOT_SET is not set or empty and can result in
│ undesired behavior
╵

╷
│ Error: Interpolated environment variable not set
│
│ The environment variable NOT_SET is not set or empty and can result in
│ undesired behavior
╵

As a result of the above problems, Terraform may not behave as intended.

Terraform v1.9.0-dev
on darwin_arm64
```

My most recent commit addresses this use case: https://github.com/hashicorp/terraform/pull/34873/commits/f20b055618f74a4893b4c4aaf58d71eec6b96afa